sky/clouds/fluidstack.py

```diff
@@ -0,0 +1,320 @@
+"""Fluidstack Cloud."""
+import json
+import requests
+import typing
+from typing import Dict, Iterator, List, Optional, Tuple
+
+from sky import clouds
+from sky import exceptions
+from sky import status_lib
+from sky.clouds import service_catalog
+
+from sky.skylet.providers.fluidstack.fluidstack_utils import (
+    FLUIDSTACK_API_KEY_PATH,
+    FLUIDSTACK_API_TOKEN_PATH,
+    FluidstackClient,
+)
+import os
+
+_CREDENTIAL_FILES = [
+    # credential files for FluidStack,
+    FLUIDSTACK_API_KEY_PATH,
+    FLUIDSTACK_API_TOKEN_PATH,
+]
+if typing.TYPE_CHECKING:
+    # Renaming to avoid shadowing variables.
+    from sky import resources as resources_lib
+
+
+@clouds.CLOUD_REGISTRY.register
+class Fluidstack(clouds.Cloud):
+    """FluidStack GPU Cloud."""
+
+    _REPR = 'Fluidstack'
+
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 57
+    # Currently, none of clouds.CloudImplementationFeatures
+    # are implemented for Fluidstack Cloud.
+    # STOP/AUTOSTOP: The Fluidstack cloud
+    # provider does not support stopping VMs.
+    _CLOUD_UNSUPPORTED_FEATURES = {
+        clouds.CloudImplementationFeatures.STOP:
+            'FluidStack cloud does not support'
+            ' stopping VMs.',
+        clouds.CloudImplementationFeatures.AUTOSTOP: 'FluidStack cloud does not'
+                                                     ' support stopping VMs.',
+        clouds.CloudImplementationFeatures.CLONE_DISK_FROM_CLUSTER:
+            'Migrating '
+            f'disk is not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.SPOT_INSTANCE:
+            'Spot instances are'
+            f' not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.CUSTOM_DISK_TIER:
+            'Custom disk tiers'
+            f' are not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.OPEN_PORTS:
+            'Opening ports'
+            f'is not supported in {_REPR}.',
+    }
```

Comment:
Should we add `clouds.CloudImplementationFeatures.MULTI_NODE`? It seems we disable it in the `node_provider`?